spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala

```diff
@@ -0,0 +1,157 @@
+      EliminateSubqueryAliases(aliasedTable) match {
+        case r @ DataSourceV2Relation(tbl: SupportsRowLevelOperations, _, _, _, _) =>
+          val operation = buildRowLevelOperation(tbl, UPDATE)
+          val table = RowLevelOperationTable(tbl, operation)
+          val rewritePlan = operation match {
+            case _: SupportsDelta =>
+              throw new AnalysisException("Delta updates are currently not supported")
+            case _ if cond.exists(SubqueryExpression.hasSubquery) =>
+              buildReplaceDataWithUnionPlan(r, table, assignments, cond)
+            case _ =>
+              buildReplaceDataPlan(r, table, assignments, cond)
+          }
+          UpdateIcebergTable(r, assignments, cond, Some(rewritePlan))
+
+        case p =>
+          throw new AnalysisException(s"$p is not an Iceberg table")
+      }
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  // if the condition does NOT contain a subquery
+  private def buildReplaceDataPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      assignments: Seq[Assignment],
+      cond: Option[Expression]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a read relation and include all required metadata columns
+    val readAttrs = dedupAttrs(relation.output ++ metadataAttrs)
+    val readRelation = relation.copy(table = table, output = readAttrs)
+
+    // build a plan with updated rows
+    val updateCond = cond.getOrElse(Literal.TrueLiteral)
+    val allRowsPlan = buildUpdateProjection(readRelation, assignments, updateCond)
+
+    // build a plan to replace read groups in the table
+    val writeRelation = relation.copy(table = table)
+    ReplaceData(writeRelation, allRowsPlan, relation)
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  // if the condition contains a subquery
+  private def buildReplaceDataWithUnionPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      assignments: Seq[Assignment],
+      cond: Option[Expression]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a read relation and include all required metadata columns
+    // the same read relation will be used to read records that must be updated and be copied over
+    // DeduplicateRelations will take care of duplicated attr IDs
+    val readAttrs = dedupAttrs(relation.output ++ metadataAttrs)
+    val readRelation = relation.copy(table = table, output = readAttrs)
+
+    // build a plan for records that match the cond and should be updated
+    val updateCond = cond.getOrElse(Literal.TrueLiteral)
```

Comment:
If the condition is empty or `true`, then this doesn't need the remaining rows plan or the union. Should we handle that case explicitly?

Reply:
I did move this to another place. Let me know if that's what you meant, @rdblue.